tokio-test/src/clock.rs

```diff
@@ -0,0 +1,256 @@
+//! A mocked clock for use with `tokio_timer` based futures.
+//!
+//! # Example
+//!
+//! ```
+//! # #[macro_use] extern crate tokio_test;
+//! # extern crate futures;
+//! # extern crate tokio_timer;
+//! # use tokio_test::clock;
+//! # use tokio_timer::Delay;
+//! # use std::time::Duration;
+//! # use futures::Future;
+//! clock::mock(|handle| {
+//!     let mut delay = Delay::new(handle.now() + Duration::from_secs(1));
+//!
+//!     assert_not_ready!(delay.poll());
+//!
+//!     handle.advance(Duration::from_secs(1));
+//!
+//!     assert_ready!(delay.poll());
+//! });
+//! ```
+
+use futures::{future::lazy, Future};
+use std::marker::PhantomData;
+use std::rc::Rc;
+use std::sync::{Arc, Mutex};
+use std::time::{Duration, Instant};
+use tokio_executor::park::{Park, Unpark};
+use tokio_timer::clock::{Clock, Now};
+use tokio_timer::Timer;
+
+/// Run the provided closure with a `MockClock` that starts at the current time.
+pub fn mock<F, R>(f: F) -> R
+where
+    F: FnOnce(&mut Handle) -> R,
+{
+    let mut mock = MockClock::new();
+    mock.enter(f)
+}
+
+/// run the provided closure witha `MockClock` that starts at the provided `Instant`.
+pub fn mock_with<F, R>(instant: Instant, f: F) -> R
+where
+    F: FnOnce(&mut Handle) -> R,
+{
+    let mut mock = MockClock::with_instant(instant);
+    mock.enter(f)
+}
+
+/// Mock clock for use with `tokio-timer` futures.
+///
+/// A mock timer that is able to advance and wake after a
+/// certain duration.
+#[derive(Debug)]
+pub struct MockClock {
+    time: MockTime,
+    clock: Clock,
+}
+
+/// A handle to the `MockClock`.
+#[derive(Debug)]
+pub struct Handle {
+    timer: Timer<MockPark>,
+    time: MockTime,
+}
+
+type Inner = Arc<Mutex<State>>;
+
+#[derive(Debug, Clone)]
+struct MockTime {
+    inner: Inner,
+    _pd: PhantomData<Rc<()>>,
+}
+
+#[derive(Debug)]
+struct MockNow {
+    inner: Inner,
+}
+
+#[derive(Debug)]
+struct MockPark {
+    inner: Inner,
+    _pd: PhantomData<Rc<()>>,
+}
+
+#[derive(Debug)]
+struct MockUnpark {
+    inner: Inner,
+}
+
+#[derive(Debug)]
+struct State {
+    base: Instant,
+    advance: Duration,
+    unparked: bool,
+    park_for: Option<Duration>,
+}
+
+impl MockClock {
+    /// Create a new `MockClock` with the current time.
+    pub fn new() -> Self {
+        MockClock::with_instant(Instant::now())
+    }
+
+    /// Create a `MockClock` that sets its current time from a duration.
+    ///
+    /// This will create a clock with `Instant::now() + duration` as the current time.
+    pub fn with_duration(duration: Duration) -> Self {
+        let instant = Instant::now() + duration;
+        MockClock::with_instant(instant)
+    }
+
+    /// Create a `MockClock` that sets its current time as the `Instant` provided.
+    pub fn with_instant(instant: Instant) -> Self {
```

Comment:
TIOLI: WDYT of calling this
```suggestion
    pub fn at_instant(instant: Instant) -> Self {
```

It's possible @carllerche won't like that since `with_` is kind of a convention for constructors, so as an alternative,
```suggestion
    pub fn with_current_time_at(instant: Instant) -> Self {
```